### PR TITLE
feat: LT-1170 - Associate users with sites

### PR DESCRIPTION
**Pull request by jfaracik**
* Add site detail view - shows site detail and users assigned to it
* Update member add page - you can now select sites to assign the user to
* Add more actions button to profile
* Add test for changing role (was previously missing)
* Add test for assigning users to sites
* Use Faker to provide fake data in E2Es
* Overhaul Members E2Es - far faster and cleaner now
* Add add_site function for creating sites in E2Es

### CODE REVIEW

File: ui_automation_tests/features/users.feature [hunk 1]
```diff
@@ -1,28 +1,20 @@
...
-    Then user is added
-    When I deactivate user then user is deactivated
-    And I reactivate user then user is reactivated
-
-  @cant_add_self @regression
-  Scenario: Can't add own user
-    Given I go to exporter homepage and choose Test Org
-    When I add self
-    Then error message is "is already a member of this organisation"
+    Then I add a member to the organisation
```
Comment (phlptt) [inline ui_automation_tests/features/users.feature]: When

File: ui_automation_tests/fixtures/register_organisation.py [hunk 2]
```diff
@@ -15,6 +15,14 @@ def register_organisation(driver, request, context, seed_data_config):
     timer.print_time("register_organisation")
 
 
+@fixture(scope="session")
+def user_details(driver, request, context, seed_data_config):
+    lite_client = get_lite_client(context, seed_data_config=seed_data_config)
+    context.exporter_headers = lite_client.context["exporter_headers"]
+    context.export_user_token = lite_client.context["export_user_token"]
+    print('user details have been set :)')
```
Comment (phlptt) [inline ui_automation_tests/fixtures/register_organisation.py]: lose print

File: ui_automation_tests/pages/member_page.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
+from shared.BasePage import BasePage
+
+
+class MemberPage(BasePage):
+
+    PANE_MORE_ACTIONS_HIDDEN_SELECTOR = "lite-more-actions__container--hidden"
```
Comment (phlptt) [inline ui_automation_tests/pages/member_page.py]: rename variable

File: ui_automation_tests/step_defs/test_members.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+from pages.add_member import AddMemberPage
+from pages.exporter_hub_page import ExporterHubPage
+from pages.member_page import MemberPage
+from pages.members_page import MembersPage
+from pages.shared import Shared
+from shared import functions
+from shared.seed_data.make_requests import post
+from shared.tools.helpers import scroll_to_element_by_id
+
+scenarios("../features/users.feature", strict_gherkin=False)
```
Comment (phlptt) [inline ui_automation_tests/step_defs/test_members.py]: can you name this members

File: ui_automation_tests/step_defs/test_members.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+fake = Faker()
+
+
+def strip_special_characters(string):
```
Comment (phlptt) [inline ui_automation_tests/step_defs/test_members.py]: should be in helpers

File: ui_automation_tests/step_defs/test_members.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    return ''.join(e for e in string if e.isalnum())
+
+
+def get_current_date_time(format_date_time=True):
```
Comment (phlptt) [inline ui_automation_tests/step_defs/test_members.py]: should be in helpers

File: ui_automation_tests/step_defs/test_members.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    date_time = datetime.now()
+    if not format_date_time:
+        return date_time
+    return strip_special_characters(str(date_time))
+
+
+def add_site(organisation_id, headers):
```
Comment (phlptt) [inline ui_automation_tests/step_defs/test_members.py]: should be somewhere else

File: ui_automation_tests/step_defs/test_members.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    data = {
```
Comment (phlptt) [inline ui_automation_tests/step_defs/test_members.py]: put in request data

File: ui_automation_tests/step_defs/test_members.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+            "city": "london",
+            "postcode": "E14GH",
+            "region": "Hertfordshire",
+            "country": "GB",
+        },
+    }
+    return post(f"/organisations/{organisation_id}/sites/", json=data, headers=headers).json()["site"]
+
+
+@then("I add a member to the organisation")
```
Comment (phlptt) [inline ui_automation_tests/step_defs/test_members.py]: when

File: ui_automation_tests/step_defs/test_members.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+    email = fake.email()
+    context.email_to_search = email
+    add_member_page.enter_email(email)
+
+    add_member_page.check_all_sites()
+    functions.click_submit(driver)
+
+
+@then("I select the member that was just added")
```
Comment (phlptt) [inline ui_automation_tests/step_defs/test_members.py]: when

File: ui_automation_tests/step_defs/test_members.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+def select_the_member_that_was_just_added(driver, context):
+    MembersPage(driver).click_view_member_link(context.email_to_search)
+
+
+@then("I deactivate them, then the member is deactivated")
```
Comment (phlptt) [inline ui_automation_tests/step_defs/test_members.py]: seperate these steps

File: ui_automation_tests/step_defs/test_members.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+
+@when("I try to deactivate myself I cannot")
+def cant_deactivate_self(driver, context):
+    exporter_hub = ExporterHubPage(driver)
+    exporter_hub.click_user_profile()
+
+    member_page = MemberPage(driver)
+    member_page.try_click_more_actions_button()
+    assert not functions.element_with_id_exists(driver, member_page.BUTTON_DEACTIVATE_ID)
```
Comment (phlptt) [inline ui_automation_tests/step_defs/test_members.py]: please add driver.set_timeout_to(0)
driver.set_timeout_to(10)

around this

File: ui_automation_tests/step_defs/test_members.py [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+    MembersPage(driver).click_view_member_link(context.email_to_search)
+
+
+@then("I deactivate them, then the member is deactivated")
+def user_deactivate(driver):
+    MemberPage(driver).click_deactivate_button()
+    assert "Deactivated" in Shared(driver).get_text_of_body(), "user status was expected to be Deactivated"
+
+
+@then("I reactivate them, then the member is reactivated")
```
Comment (phlptt) [inline ui_automation_tests/step_defs/test_members.py]: seperate these steps

File: ui_automation_tests/features/members.feature [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+    And I add a member to the organisation
+    And I select the member that was just added
+    # Edit the user
+    And I deactivate them, then the member is deactivated
+    And I reactivate them, then the member is reactivated
+    And I change what sites they're assigned to
+    Then I change their role
+
+  @reactivate_oneself @regression
+  Scenario: Reactivate oneself
```
Comment (phlptt) [inline ui_automation_tests/features/members.feature]: change title

File: ui_automation_tests/features/sites.feature [hunk 15]
```diff
@@ -19,7 +19,8 @@ Feature: I  want to add and rename my sites
     Given I go to exporter homepage and choose Test Org
     When I click on the manage my organisation link
     When I click sites link
```
Comment (phlptt) [inline ui_automation_tests/features/sites.feature]: change to and

File: sites/views.py [hunk 16]
```diff
@@ -58,21 +58,42 @@ def post(self, request, **kwargs):
         return redirect(reverse_lazy("sites:sites"))
 
 
+class ViewSite(TemplateView):
+    def get(self, request, *args, **kwargs):
+        organisation_id = str(request.user.organisation)
+        site = get_site(request, organisation_id, kwargs["pk"])
+        organisation = get_organisation(request, organisation_id)
+        # user_permissions = get_user_permissions(request)
```
Comment (Mark-Spiers-kainos) [inline sites/views.py]: a lot commented out here without any reason attached?

File: Pipfile [hunk 17]
```diff
@@ -13,6 +13,7 @@ pytest-html = "*"
 prospector = "==1.1.6.2"
 pytest-xdist = "*"
 black = "==19.10b0"
+faker = "*"
```
Comment (BenAndrew-Kainos) [inline Pipfile]: What is faker?

File: sites/forms.py [hunk 18]
```diff
@@ -20,13 +20,13 @@ def new_site_form():
     )
 
 
-def edit_site_form(title):
+def edit_site_form(site):
     return Form(
-        title=title,
+        title="Edit " + site["name"],
```
Comment (BenAndrew-Kainos) [inline sites/forms.py]: Move the Edit to strings

File: sites/forms.py [hunk 19]
```diff
@@ -20,13 +20,13 @@ def new_site_form():
...
         questions=[
             TextInput(title="Name of site", name="name"),
             Heading("Where is the site based?", HeadingStyle.M),
             *address_questions(get_countries(None, True)),
         ],
-        back_link=BackLink("Back to Sites", reverse_lazy("sites:sites")),
+        back_link=BackLink("Back to " + site["name"], reverse_lazy("sites:site", kwargs={"pk": site["id"]})),
```
Comment (BenAndrew-Kainos) [inline sites/forms.py]: Move the Back to to strings

File: sites/services.py [hunk 20]
```diff
@@ -13,9 +13,9 @@ def get_sites(request, organisation_id, convert_to_options=False):
         sites_options = []
 
         for site in data:
-            primary_site = "(Primary Site)" if site.get("id") == organisation["primary_site"]["id"] else ""
+            primary_site = " (Primary Site)" if site.get("id") == organisation["primary_site"]["id"] else ""
```
Comment (BenAndrew-Kainos) [inline sites/services.py]: Please move `" (Primary Site)"` to strings

File: sites/services.py [hunk 21]
```diff
@@ -51,3 +51,9 @@ def put_site(request, organisation_id, pk, json):
 def post_sites(request, organisation_id, json):
     data = post(request, ORGANISATIONS_URL + organisation_id + SITES_URL, json)
     return data.json(), data.status_code
+
+
+def put_assign_sites(request, pk, json):
+    # TODO stop hardcoding sites!
+    data = put(request, USERS_URL + str(pk) + "/sites/", json)
```
Comment (BenAndrew-Kainos) [inline sites/services.py]: Could you move `/sites/`to constants

File: Pipfile (hunk 17 above)
Comment (BenAndrew-Kainos) [reply Pipfile]: Oh just seen its use. That's pretty cool

File: users/forms.py [hunk 22]
```diff
@@ -17,6 +18,11 @@ def add_user_form(request):
                 title=strings.USER_ROLE_QUESTION,
                 include_default_select=False,
             ),
+            Checkboxes(
+                title="What sites should the user be assigned to?",
```
Comment (BenAndrew-Kainos) [inline users/forms.py]: Move to strings!

File: users/forms.py [hunk 23]
```diff
@@ -39,3 +45,12 @@ def edit_user_form(request, user_id, can_edit_role: bool):
         back_link=BackLink(strings.USER_EDIT_FORM_BACK_TO_USER, reverse_lazy("users:user", kwargs={"pk": user_id})),
         default_button_name=strings.USER_EDIT_FORM_SAVE,
     )
+
+
+def assign_sites(request):
+    return Form(
+        title="Assign user to sites",
+        description="Select all sites that apply.",
```
Comment (BenAndrew-Kainos) [inline users/forms.py]: Move to strings!

File: conf/decorators.py [hunk 24]
```diff
@@ -0,0 +1,26 @@
+from django.utils.functional import wraps
+
+from conf.exceptions import PermissionDeniedError
+from core import helpers
+
+
+def has_permission(permission):
```
Comment (a-bek) [inline conf/decorators.py]: Should all these permission utilites be put into their own module?  (including the ones below inside helpers.py) Easier to find

File: users/views.py [hunk 25]
```diff
@@ -47,19 +48,19 @@ def init(self, request, **kwargs):
...
-        user = get_user(request)
-        is_user_super_user = is_super_user(user)
-
         request_user = get_organisation_user(request, str(request.user.organisation), str(kwargs["pk"]))
+        user = get_user(request)
         is_request_user_super_user = is_super_user(request_user)
+        is_user_super_user = is_super_user(user)
 
-        show_change_status = not is_request_user_super_user
+        show_change_status = is_user_super_user and not is_request_user_super_user and user["id"] != request_user["id"]
```
Comment (a-bek) [inline users/views.py]: Maybe assign `is_self_editing = user["id"] != request_user["id"]`

```
is_self_editing = user["id"] == request_user["id"]

show_change_status = not is_self_editing and is_user_super_user
show_change_role = not is_self_editing and is_user_super_user
show_assign_sites = not is_self_editing and not is_request_user_super_user
```

Think that might work!

File: conf/decorators.py (hunk 24 above)
Comment (jfaracik) [reply conf/decorators.py]: I like them in their respective places truthfully, but worthy of a discussion for sure!

**Pull request merged**